### PR TITLE
Search clean-up: consolidate Filter specifications

### PR DESCRIPTION
**Pull request by crisptrutski**
Consolidate a bunch of maps and special cases into a central specification.

### CODE REVIEW

File: src/metabase/search/config.clj [hunk 1]
```diff
@@ -90,6 +91,54 @@
...
+  [:map {:closed true}
+   [:key                               :keyword]
+   [:type                              :keyword]
+   [:field            {:optional true} :string]
+   [:context-key      {:optional true} :keyword]
+   [:supported-value? {:optional true} ifn?]
+   [:required-feature {:optional true} :keyword]])
+
+(def ^:private Filter
+  "Map with the various allowed search parameters, used to construct the SQL query."
```
[the remark this answers is not in the document]
Comment (crisptrutski) [reply src/metabase/search/config.clj:105]: To make matters worse...

```
(def SearchContext
  "Map with the various allowed search parameters, used to construct the SQL query."
```

File: src/metabase/search/config.clj [hunk 2]
```diff
@@ -90,6 +91,54 @@
    :view-count          2
    :text                10})
 
+(def ^:private FilterSpec
+  "Map with the various allowed search parameters, used to construct the SQL query."
```
Comment (crisptrutski) [inline src/metabase/search/config.clj:95]:
```suggestion
```

File: src/metabase/search/config.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (crisptrutski) [reply src/metabase/search/config.clj:105]:
```suggestion
```

[the remark this answers is not in the document]
Comment (crisptrutski) [reply src/metabase/search/config.clj]: Hmm, I should probably rename them too.

How about `FilterUserDefinition` and `Filter`?

[the remark this answers is not in the document]
Comment (piranha) [reply src/metabase/search/config.clj]: More understandable, yes

**Review by sansolovyov — APPROVED** [review]
soooo good

**Pull request merged**